path_functions.cpp

```diff
@@ -4,104 +4,159 @@
+#ifndef CASE_SENSITIVE_FILESYSTEM
+  return char_path + "/emotions/" + p_file;
+#else
+  return get_case_sensitive_path(
+         get_case_sensitive_path(char_path) + "/emotions/" + p_file);
+#endif
 }
-QString AOApplication::get_sounds_path()
+
+QString AOApplication::get_sounds_path(QString p_file)
 {
-  return get_base_path() + "sounds/general/";
+  QString path = get_base_path() + "sounds/general/" + p_file;
+#ifndef CASE_SENSITIVE_FILESYSTEM
+  return path;
+#else
+  return get_case_sensitive_path(path);
+#endif
 }
+
 QString AOApplication::get_music_path(QString p_song)
 {
-  return get_base_path() + "sounds/music/" + p_song.toLower();
+  QString path = get_base_path() + "sounds/music/" + p_song;
+#ifndef CASE_SENSITIVE_FILESYSTEM
+  return path;
+#else
+  return get_case_sensitive_path(path);
+#endif
 }
 
-QString AOApplication::get_background_path()
+QString AOApplication::get_background_path(QString p_file)
 {
-  if (courtroom_constructed)
-    return w_courtroom->get_background_path();
+  QString bg_path = get_base_path() + "background/" + w_courtroom->get_current_background();
+  if (courtroom_constructed) {
+#ifndef CASE_SENSITIVE_FILESYSTEM
+    return bg_path + "/" + p_file;
+#else
+    return get_case_sensitive_path(
+           get_case_sensitive_path(bg_path) + "/" + p_file);
+#endif
+  }
   //this function being called when the courtroom isn't constructed makes no sense
   return "";
 }
 
-QString AOApplication::get_default_background_path()
+QString AOApplication::get_default_background_path(QString p_file)
 {
-  return get_base_path() + "background/default/";
+  QString path = get_base_path() + "background/default/" + p_file;
+#ifndef CASE_SENSITIVE_FILESYSTEM
+  return path;
+#else
+  return get_case_sensitive_path(path);
+#endif
 }
 
-QString AOApplication::get_evidence_path()
+QString AOApplication::get_evidence_path(QString p_file)
 {
-    QString default_path = "evidence/";
-    QString alt_path = "items/";
-    if (dir_exists(default_path))
-      return get_base_path() + default_path;
-    else if (dir_exists(alt_path))
-      return get_base_path() + alt_path;
-    else
-      return get_base_path() + default_path;
+  QString path = get_base_path() + "evidence/" + p_file;
+#ifndef CASE_SENSITIVE_FILESYSTEM
+  return path;
+#else
+  return get_case_sensitive_path(path);
+#endif
 }
 
-QString Courtroom::get_background_path()
-{
-  return ao_app->get_base_path() + "background/" + current_background.toLower() + "/";
-}
+QString AOApplication::get_case_sensitive_path(QString p_file) {
+  QFileInfo file(p_file);
 
-QString Courtroom::get_default_background_path()
-{
-  return ao_app->get_base_path() + "background/default/";
+  //quick check to see if it's actually there first
+  if (file.exists()) return p_file;
+
+  QString file_name = file.fileName();
+  QString file_path = file.absolutePath();
+
+  QRegExp file_rx = QRegExp(file_name, Qt::CaseInsensitive);
```

Comment:
Does this work if the directory is case-insensitive as well?

Reply:
it assumes that the parent of p_file definitely exists (with proper case) and does not differentiate between directory and file. so for example, for the file:

base/characters/SoMeThInG/emotions/BUTTON1_OFF.PNG

this need to be called twice; the first call would be get_case_sensitive_path("base/characters/SoMeThInG") -> return "base/characters/Something"(the file that actually exists), and the second call would be
get_case_sensitive_path("base/characters/Something/emotions/BUTTON1_OFF.PNG") -> return "base/characters/Something/emotions/button1_off.png"(again, a file that actually exists)

the solution is sound and works on Ubuntu 18.10 with ext4 filesystem. ive yet to test it with cyrillic characters, but assuming Qt is smart with encodings it should work

Reply:
so, it's not recursive by default? the issue is that some asset authors have taken it upon themselves to organize their emotes by hacking their own directories into char.ini files. Those should be checked as well.

Reply:
nope, I did not thoroughly consider that case (honestly hoped it had fallen out of fashion by now). I suppose it's not too much of a performance issue to check for existence of parent folder though. ill make it recursive